File: examples/hello.m
:- module hello.
:- interface.
:- import_module int.

    % Phase A hello-world for the Mercury bridge.
    %
    % Mercury's standard `:- pred main(io::di, io::uo) is det.` entry point
    % requires io.write_string from Mercury's stdlib io module, which the
    % MercuryBridge does not yet shim — see ROADMAP →
    % BRIDGE_mercury_strings.
    %
    % This file uses the `:- pred main_int(int::out) is det.` convention
    % that MercuryBridge synthesises into a Frankenstein-Core main
    % returning Int.  Standalone Mercury programs with single-clause
    % fact bodies (e.g. `main_int(13).`) now compile cleanly thanks to
    % HldsParse stripping the trailing `.` from goal text so the literal
    % `13` is recognised by translateGoalK's bind-to-literal case.
    %
    % Expected output: 13   (byte-length of "Hello, World!" — hardcoded
    %                        since string printing isn't wired up yet)
:- pred main_int(int::out) is det.

:- implementation.
main_int(13).

File: examples/hello.rs
// Phase A hello-world for the Rust bridge — degraded form.
//
// The Rust bridge does not yet shim std::io / core::fmt::Arguments, so
// `println!("Hello, World!")` still fails to link (Arguments::new,
// std::io::_print remain unresolved).  See ROADMAP →
// BRIDGE_rust_strings.
//
// This degraded form computes the byte-length of "Hello, World!" via
// `str::len`, returning the i64 through the proven main-returns-Int
// wrapper.  The Rust bridge now remaps `core::str::<impl str>::len` to
// the runtime's `kk_str_len`, so this compiles and runs end-to-end.
//
// Expected output: 13   (byte-length of "Hello, World!")

#[no_mangle]
pub fn main() -> i64 {
    "Hello, World!".len() as i64
}
